File: apina-core/src/test/java/fi/evident/apina/java/reader/ReflectionClassMetadataLoader.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.JavaModel;

import java.lang.annotation.Annotation;

import static fi.evident.apina.java.reader.ClassReaderUtils.loadClass;

public final class ReflectionClassMetadataLoader {

    public static JavaModel loadClassesFromInheritanceTree(Class<?> clazz) {
        JavaModel result = new JavaModel();

        for (Class<?> cl = clazz; cl != null && cl != Object.class; cl = cl.getSuperclass())
            loadClassesAt(result, cl);

        return result;
    }

    private static void loadClassesAt(JavaModel result, Class<?> cl) {
        if (!result.containsClass(cl.getName())) {
            result.addClass(loadClass(cl));

            for (Annotation annotation : cl.getDeclaredAnnotations())
                loadClassesAt(result, annotation.annotationType());

            for (Class<?> interfaceClass : cl.getInterfaces())
                loadClassesAt(result, interfaceClass);
        }
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/java/reader/ReflectionClassMetadataLoader.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.model.JavaModel

import fi.evident.apina.java.reader.ClassReaderUtils.loadClass

inline fun <reified T : Any> JavaModel.loadClassesFromInheritanceTree() {
    loadClassesFromInheritanceTree(T::class.java)
}

fun JavaModel.loadClassesFromInheritanceTree(clazz: Class<*>) {
    var cl: Class<*>? = clazz
    while (cl != null && cl != Any::class.java) {
        loadClassesAt(cl)
        cl = cl.superclass
    }
}

private fun JavaModel.loadClassesAt(cl: Class<*>) {
    if (!containsClass(cl.name)) {
        addClass(loadClass(cl))

        for (annotation in cl.declaredAnnotations)
            loadClassesAt(annotation.annotationClass.java)

        for (interfaceClass in cl.interfaces)
            loadClassesAt(interfaceClass)
    }
}

